developer_docs/reorganizations.md

```diff
@@ -0,0 +1,51 @@
+# How different parts of the state handle reorganizations
+Neptune is a blockchain which features recursive STARK proofs as part of its
+consensus mechanism. This implies that participants can synchronize trustlessly
+by simply downloading the latest block and verifying this. Unlike most other
+blockchains, it is not necessary to download all historical blocks to get a
+cryptographically verified view of the state of the blockchain.
+
+It is possible, though, to run an archival node that downloads all historical
+blocks. This archival node comes with additional functionality such as being
+able to reconstruct transaction's membership proofs, provide some historical
+transaction statistics, and allow other archival nodes to synchronize.
+
+This document provides an overview of how different parts of the client's state
+handle reorganizations.
+
+## State overview
+The client's state consists of the following parts:
+- wallet
+- light state
+- archival state (optional)
+- mempool
+
+The wallet handles transactions that the client holds the spending keys for.
+The light state contains the latest blocks which verifies the validity of the
+entire history of the blockchain. The archival state is optional and allows,
+among other things, the client to re-synchronize wallets that are no longer
+up-to-date. The mempool keeps track of transactions that are not yet included
+in blocks, thus allowing miners to confirm transactions by picking some from
+the mempool to include in the next block.
+
+### Wallet
+The wallet can handle reorganizations that are up to `n` blocks deep, where `n`
```

Comment:
What happens if my wallet has a (monitored) utxo that gets dropped from the mempool during re-org?    iiuc, it seems the wallet would be out-of-sync with the mempool, but not the wallet's fault.

Reply:
The wallet currently *only* deals with and displays transactions that have been mined. It completely ignores anything that's in the mempool.

Reply:
Does that address your question? If not, please elaborate :)

Reply:
> It completely ignores anything that's in the mempool.

Ok, so that implies that spends are not reflected in the wallet prior to confirmation either.

Eg, when I send a payment P for amount A, and P is not yet confirmed in a block:
1. A is not reflected in the wallet balance.
2. P is not shown in the wallet's transaction history.
3. The wallet would let me attempt to re-spend P's inputs because it is unaware that those utxo are already in the mempool.   But any such attempt would fail, because spend() actually does check the mempool.

Is the above accurate?

If so, then I think this behavior deserves a mention in the docs somewhere also, until such time as the wallet becomes mempool-aware.